// metro/src/commands/build.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true,
});
exports.default = void 0;
var _cliUtils = require("../cli-utils");
var _parseKeyValueParamArray = _interopRequireDefault(
  require("../cli/parseKeyValueParamArray"),
);
var _TerminalReporter = _interopRequireDefault(
  require("../lib/TerminalReporter"),
);
var _metroConfig = require("metro-config");
var _metroCore = require("metro-core");
function _interopRequireDefault(e) {
  return e && e.__esModule ? e : { default: e };
}
const term = new _metroCore.Terminal(process.stdout);
const updateReporter = new _TerminalReporter.default(term);
var _default = () => ({
  command: "build <entry>",
  desc: "Generates a JavaScript bundle containing the specified entrypoint and its descendants",
  builder: (yargs) => {
    yargs.option("project-roots", {
      alias: "P",
      type: "string",
      array: true,
    });
    yargs.option("out", {
      alias: "O",
      type: "string",
      demandOption: true,
    });
    yargs.option("platform", {
      alias: "p",
      type: "string",
    });
    yargs.option("output-type", {
      alias: "t",
      type: "string",
    });
    yargs.option("max-workers", {
      alias: "j",
      type: "number",
    });
    yargs.option("minify", {
      alias: "z",
      type: "boolean",
    });
    yargs.option("dev", {
      alias: "g",
      type: "boolean",
    });
    yargs.option("source-map", {
      type: "boolean",
    });
    yargs.option("source-map-url", {
      type: "string",
    });
    yargs.option("legacy-bundler", {
      type: "boolean",
    });
    yargs.option("config", {
      alias: "c",
      type: "string",
    });
    yargs.option("transform-option", {
      type: "string",
      array: true,
      alias: "transformer-option",
      coerce: _parseKeyValueParamArray.default,
      describe:
        "Custom transform options of the form key=value. URL-encod
